bootstrap/cmd/bootstrap/app/kfctlUtil.go

```diff
@@ -0,0 +1,92 @@
+package app
+
+import (
+	kftypes "github.com/kubeflow/kubeflow/bootstrap/pkg/apis/apps"
+	kstypes "github.com/kubeflow/kubeflow/bootstrap/pkg/apis/apps/kfdef/v1alpha1"
+	kfdefs "github.com/kubeflow/kubeflow/bootstrap/pkg/apis/apps/kfdef/v1alpha1"
+	"github.com/kubeflow/kubeflow/bootstrap/pkg/kfapp/gcp"
+	"github.com/kubeflow/kubeflow/bootstrap/pkg/kfapp/ksonnet"
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"os"
+	"path"
+	"path/filepath"
+)
+
+func (cr *CreateRequest) ToKfdef(appDir string, repo string, istio bool) (*kfdefs.KfDef, error) {
+	kfDef := &kfdefs.KfDef{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "KfDef",
+			APIVersion: "kfdef.apps.kubeflow.org/v1alpha1",
+		},
+		Spec: kfdefs.KfDefSpec{},
+	}
+	kfDef.Spec.ComponentConfig = cr.AppConfig
+	kfDef.Name = cr.Name
+	kfDef.Spec.AppDir = appDir
+	kfDef.Spec.Platform = kftypes.GCP
+	kfDef.Namespace = cr.Namespace
+	kfDef.Spec.Version = cr.Version
+	kfDef.Spec.Repo = repo
+	kfDef.Spec.Project = cr.Project
+	// TODO
+	//kfDef.Spec.UseBasicAuth = options[string(kftypes.USE_BASIC_AUTH)].(bool)
+	kfDef.Spec.UseIstio = istio
+	return kfDef, nil
+}
+
+func (s *ksServer) DeployWithKfctl(req *CreateRequest) error{
+	// pull versioned kubeflow repo
+	versionedRegPath, err := s.getRegistryUri(&kstypes.RegistryConfig{
+		Name: "kubeflow",
+		Repo: "https://github.com/kubeflow/kubeflow.git",
+		Version: req.Version,
+		Path: "kubeflow",
+	})
+	if err != nil {
+		return err
+	}
+	versionedKfRepo := filepath.Dir(versionedRegPath)
+
+	// pull source repo to random tmp dir
+	repoDir, err := s.CloneRepoToLocal(req.Project, req.Token, GetRepoNameKfctl(req.Project))
+	// clean up tmp dir afterwards
+	defer os.RemoveAll(repoDir)
+	if err != nil {
+		return err
+	}
+	ksAppDir := path.Join(repoDir, GetRepoNameKfctl(req.Project), req.Version, req.Name)
+	kfdef, err := req.ToKfdef(ksAppDir, versionedKfRepo, s.installIstio)
+	if err != nil {
+		return err
+	}
+	ts := oauth2.StaticTokenSource(&oauth2.Token{
+		AccessToken: req.Token,
+	})
+	ctx := context.Background()
+
+	// run gcp generate / apply
+	gcpApp := gcp.BuildKfApp(kfdef, oauth2.NewClient(ctx, ts), ts)
```

Comment:
it might be easier to use coordinator which wraps the platform and package managers within its Generate and Apply. That way it would also switch package managers based on whats in the [KfDefSpec](https://github.com/kkasravi/kubeflow/blob/kfctl_switch_packagemanagers/bootstrap/pkg/apis/apps/kfdef/v1alpha1/application_types.go#L39)

Reply:
Agree, the only problem is now in coordinator
```platform.apply()``` and ```pkgManager.apply()``` are executed together and no space to insert credentials between them.

How about change coordinator order to be:
```
platform.generate()
platform.apply()
pkgManager.generate()
pkgManager.apply()
```
?

Reply:
then coordinator would not implement KfApp, I think that may have a number of impacts.

It sounds like platform.Apply get credentials that PkgManager.Apply needs?
What if
-  Coordinator has a platformPostHook callback so we don't surface GCP in Coordinator?
or
- the platform is passed to the package manager and the package manager can call the platforms applyCallback?

Reply:
@kunmingg @kkasravi Don't Apply and Generate take an enum specifying what phase to run e.g. platform and k8s so can't you do

coordinator.generate(platform)
coordinator.apply(platform)
coordinator.generate(k8s)
coordinator.apply(k8s)

Would this hide the details of the package manager? It would be nice if the app didn't have to change when the coordinator logic changed; e.g. to kustomize.